src/modules/schnorr_adaptor/main_impl.h

```diff
@@ -0,0 +1,286 @@
+
+    /* R0 */
+    ret &= !!secp256k1_xonly_pubkey_parse(ctx, &pkr0, &sig65[1]);
+    ret &= !!secp256k1_xonly_pubkey_load(ctx, &r0, &pkr0);
+
+    /* Compute e */
+    secp256k1_fe_get_b32(buf, &pk.x);
+    secp256k1_schnorrsig_challenge(&e, &sig65[1], msg32, msglen, buf);
+
+    /* Compute rj = s0*G + (-e) * pkj */
+    secp256k1_scalar_negate(&e, &e);
+    secp256k1_gej_set_ge(&pkj, &pk);
+    secp256k1_ecmult(&rj, &pkj, &e, &s0);
+
+    /* R */
+    secp256k1_ge_set_gej_var(&r, &rj);
+    ret &= !secp256k1_ge_is_infinity(&r);
+
+    /* T = R0 + (- R) */
+    secp256k1_gej_neg(&rj, &rj);
+    secp256k1_gej_add_ge_var(&tj, &rj, &r0, NULL);
+    if (sig65[0] == SECP256K1_TAG_PUBKEY_EVEN) {
+        ;
+    } else if (sig65[0] == SECP256K1_TAG_PUBKEY_ODD) {
+        secp256k1_gej_neg(&tj, &tj);
+    } else {
+        ret = 0;
+    }
+    secp256k1_ge_set_gej(&t, &tj);
+    ret &= !!secp256k1_eckey_pubkey_serialize(&t, t33, &size, 1);
+
+    secp256k1_memczero(t33, 33, !ret);
+    secp256k1_scalar_clear(&s0);
+
+    return ret;
+}
+
+int secp256k1_schnorr_adaptor_adapt(const secp256k1_context *ctx, unsigned char *sig64, const unsigned char *sig65, const unsigned char *t32) {
+    secp256k1_scalar s0;
+    secp256k1_scalar s;
+    secp256k1_scalar t;
+    int overflow;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(sig65 != NULL);
+    ARG_CHECK(t32 != NULL);
+
+    /* s0 */
+    secp256k1_scalar_set_b32(&s0, &sig65[33], &overflow);
+    ret &= !overflow;
+
+    /* t */
+    secp256k1_scalar_set_b32(&t, t32, &overflow);
+    ret &= !overflow;
+
+    if (sig65[0] == SECP256K1_TAG_PUBKEY_EVEN) {
+        secp256k1_scalar_add(&s, &s0, &t);
+    } else if (sig65[0] == SECP256K1_TAG_PUBKEY_ODD) {
+        secp256k1_scalar_negate(&t, &t);
+        secp256k1_scalar_add(&s, &s0, &t);
+    } else {
+        ret = 0;
+    }
+
+    memset(sig64, 0, 64);
+    memcpy(sig64, &sig65[1], 32);
+    secp256k1_scalar_get_b32(&sig64[32], &s);
+
+    return ret;
+}
+
+int secp256k1_schnorr_adaptor_extract_adaptor(const secp256k1_context *ctx, unsigned char *t32, const unsigned char *sig65, const unsigned char *sig64) {
+    secp256k1_scalar s0;
+    secp256k1_scalar s;
+    secp256k1_scalar t;
+    int overflow;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(t32 != NULL);
+    ARG_CHECK(sig65 != NULL);
+    ARG_CHECK(sig64 != NULL);
+
+    /* s0 */
+    secp256k1_scalar_set_b32(&s0, &sig65[33], &overflow);
+    ret &= !overflow;
+
+    /* s */
+    secp256k1_scalar_set_b32(&s, &sig64[32], &overflow);
+    ret &= !overflow;
+
+    if (sig65[0] == SECP256K1_TAG_PUBKEY_EVEN) {
+        secp256k1_scalar_negate(&s0, &s0);
+        secp256k1_scalar_add(&t, &s, &s0);
+    } else if (sig65[0] == SECP256K1_TAG_PUBKEY_ODD) {
+        secp256k1_scalar_negate(&s, &s);
+        secp256k1_scalar_add(&t, &s0, &s);
+    } else {
+        ret = 0;
+    }
+
+    secp256k1_scalar_get_b32(t32, &t);
+
```

Comment:
We need to clear the scalars `s` and `t`. And conditionally set `t32` to `0` if `ret == 0`.

Reply:
Without this, if the user passes an overflowed `s`, and `s0`, this API generates a valid `t`, but we need to set `t` to `0`.